sdks/go/pkg/beam/util/starcgenx/starcgenx.go

```diff
@@ -0,0 +1,565 @@
+	case *types.Func:
+		sig := obj.Type().(*types.Signature)
+		if recv := sig.Recv(); recv != nil {
+			// Methods don't need registering, but they do need shim generation.
+			e.Printf("%s: %q is a method of %v -> %v--- %T %v %v %v\n",
+				fset.Position(id.Pos()), id.Name, recv.Type(), obj, obj, id, obj.Pkg(), obj.Type())
+			if !lifecycleMethodName(id.Name) {
+				// If this is not a lifecycle method, we should ignore it.
+				return
+			}
+		} else if id.Name != "init" {
+			// init functions are special and should be ignored.
+			// Functions need registering, as well as shim generation.
+			e.Printf("%s: %q is a top level func %v --- %T %v %v %v\n",
+				fset.Position(id.Pos()), ident, obj, obj, id, obj.Pkg(), obj.Type())
+			e.functions[ident] = struct{}{}
+		}
+		// For functions from other packages.
+		if pkg.Name() != e.Package {
+			e.imports[pkg.Path()] = struct{}{}
+		}
+
+		e.funcs[e.sigKey(sig)] = sig
+		e.extractParameters(sig)
+		e.Printf("\t%v\n", sig)
+	case *types.TypeName:
+		e.Printf("%s: %q is a type %v --- %T %v %v %v %v\n",
+			fset.Position(id.Pos()), id.Name, obj, obj, id, obj.Pkg(), obj.Type(), obj.Name())
+		// Probably need to sanity check that this type actually is/has a ProcessElement
+		// or MergeAccumulators defined for this type so unnecessary registrations don't happen,
+		// an can explicitly produce an error if an explicitly named type *isn't* a DoFn or CombineFn.
+		e.extractType(ot)
+	default:
+		e.Printf("%s: %q defines %v --- %T %v %v %v\n",
+			fset.Position(id.Pos()), types.ObjectString(obj, e.qualifier), obj, obj, id, obj.Pkg(), obj.Type())
+	}
+}
+
+func (e *Extractor) extractType(ot *types.TypeName) {
+	name := types.TypeString(ot.Type(), e.qualifier)
+	// Unwrap an alias by one level.
+	// Attempting to deference a full chain of aliases runs the risk of crossing
+	// a visibility boundary such as internal packages.
+	// A single level is safe since the code we're analysing imports it,
+	// so we can assume the generated code can access it too.
+	if ot.IsAlias() {
+		if t, ok := ot.Type().(*types.Named); ok {
+			ot = t.Obj()
+			name = types.TypeString(t, e.qualifier)
+
+			if pkg := t.Obj().Pkg(); pkg != nil {
+				e.imports[pkg.Path()] = struct{}{}
+			}
+		}
+	}
+	e.types[name] = struct{}{}
+}
+
+// Examines the signature and extracts types of parameters for generating
+// necessary imports and emitter and iterator code.
+func (e *Extractor) extractParameters(sig *types.Signature) {
+	in := sig.Params() // *types.Tuple
+	for i := 0; i < in.Len(); i++ {
+		s := in.At(i) // *types.Var
+
+		// Pointer types need to be iteratively unwrapped until we're at the base type,
+		// so we can get the import if necessary.
+		t := s.Type()
+		p, ok := t.(*types.Pointer)
+		for ok {
+			t = p.Elem()
+			p, ok = t.(*types.Pointer)
+		}
+		// Here's were we ensure we register new imports.
+		if t, ok := t.(*types.Named); ok {
+			if pkg := t.Obj().Pkg(); pkg != nil {
+				e.imports[pkg.Path()] = struct{}{}
```

Comment:
For my education:
1. What is `struct{}{}` ? Is it an empty struct, initialized to be empty?
2. What is the point of setting this here. Is it just a matter of anything not nil?

Reply:
1. Exactly, empty struct, being initialized. 
2. an empty struct has size 0, so this makes maps functionally a set, with only the map bits. 
This is useful for de-duplicating entries by key.

The alternative is to use booleans, which is more concise when subsequently checking presence
eg.
foo := make(map[string]bool)
if foo["bar] { 
 Baz()
}

vs  
foo := make(map[string]struct{})
if _, ok := foo["bar"]; ok {
  Baz()
}

Reply:
Thanks!